File: dart/lib/src/resources.dart
import 'dart:convert';
import 'client.dart';
import 'models.dart';

// ── RecommendationsResource ───────────────────────────────────────────────────

class RecommendationsResource {
  final RecommendAIClient _client;
  const RecommendationsResource(this._client);

  Future<List<Recommendation>> get(String userId, {int limit = 10}) async {
    final resp = await _client.get('/api/recommendations',
        queryParams: {'user_id': userId, 'limit': '$limit'});
    final body = jsonDecode(resp.body) as Map<String, dynamic>;
    final list = body['recommendations'] as List<dynamic>;
    return list
        .cast<Map<String, dynamic>>()
        .map(Recommendation.fromJson)
        .toList();
  }

  Future<List<Recommendation>> similar(String itemId, {int limit = 10}) async {
    final resp = await _client.get(
        '/api/recommendations/similar/$itemId',
        queryParams: {'limit': '$limit'});
    final body = jsonDecode(resp.body) as Map<String, dynamic>;
    final list = body['recommendations'] as List<dynamic>;
    return list
        .cast<Map<String, dynamic>>()
        .map(Recommendation.fromJson)
        .toList();
  }

  Future<List<Recommendation>> popular({int limit = 10, String? category}) async {
    final params = <String, String>{'limit': '$limit'};
    if (category != null) params['category'] = category;
    final resp = await _client.get('/api/recommendations/popular',
        queryParams: params);
    final body = jsonDecode(resp.body) as Map<String, dynamic>;
    final list = body['recommendations'] as List<dynamic>;
    return list
        .cast<Map<String, dynamic>>()
        .map(Recommendation.fromJson)
        .toList();
  }
}

// ── UsersResource ─────────────────────────────────────────────────────────────

class UsersResource {
  final RecommendAIClient _client;
  const UsersResource(this._client);

  Future<User> create(String userId,
      {Map<String, dynamic> properties = const {}}) async {
    final resp = await _client.post('/api/users', {
      'user_id':    userId,
      'properties': properties,
    });
    return User.fromJson(jsonDecode(resp.body) as Map<String, dynamic>);
  }

  Future<User> getUser(String userId) async {
    final resp = await _client.get('/api/users/$userId');
    return User.fromJson(jsonDecode(resp.body) as Map<String, dynamic>);
  }

  Future<User> update(String userId, Map<String, dynamic> properties) async {
    final resp = await _client.put('/api/users/$userId', {
      'properties': properties,
    });
    return User.fromJson(jsonDecode(resp.body) as Map<String, dynamic>);
  }

  Future<void> delete(String userId) => _client.delete('/api/users/$userId');
}

// ── ItemsResource ─────────────────────────────────────────────────────────────

class ItemsResource {
  final RecommendAIClient _client;
  const ItemsResource(this._client);

  Future<Item> create(String itemId,
      {Map<String, dynamic> properties = const {}}) async {
    final resp = await _client.post('/api/items', {
      'item_id':    itemId,
      'properties': properties,
    });
    return Item.fromJson(jsonDecode(resp.body) as Map<String, dynamic>);
  }

  Future<Item> getItem(String itemId) async {
    final resp = await _client.get('/api/items/$itemId');
    return Item.fromJson(jsonDecode(resp.body) as Map<String, dynamic>);
  }

  Future<Item> update(String itemId, Map<String, dynamic> properties) async {
    final resp = await _client.put('/api/items/$itemId', {
      'properties': properties,
    });
    return Item.fromJson(jsonDecode(resp.body) as Map<String, dynamic>);
  }

  Future<void> delete(String itemId) => _client.delete('/api/items/$itemId');

  Future<List<Item>> upsert(List<Map<String, dynamic>> items) async {
    final resp = await _client.post('/api/items/bulk', {'items': items});
    final body = jsonDecode(resp.body) as Map<String, dynamic>;
    final list = body['items'] as List<dynamic>;
    return list.cast<Map<String, dynamic>>().map(Item.fromJson).toList();
  }
}

// ── InteractionsResource ──────────────────────────────────────────────────────

class InteractionsResource {
  final RecommendAIClient _client;
  const InteractionsResource(this._client);

  Future<Interaction> create(
    String userId,
    String itemId,
    InteractionType interactionType, {
    double? value,
    Map<String, dynamic>? metadata,
  }) async {
    final body = <String, dynamic>{
      'user_id':          userId,
      'item_id':          itemId,
      'interaction_type': interactionType.name,
    };
    if (value    != null) body['value']    = value;
    if (metadata != null) body['metadata'] = metadata;
    final resp = await _client.post('/api/interactions', body);
    return Interaction.fromJson(jsonDecode(resp.body) as Map<String, dynamic>);
  }
}

File: rust/src/resources.rs
use std::collections::HashMap;

use serde::Serialize;
use serde_json::Value;

use crate::client::RecommendAIClient;
use crate::errors::Error;
use crate::models::{Interaction, InteractionType, Item, Recommendation, RecommendationsResponse, User};

// ── RecommendationsResource ───────────────────────────────────────────────────

pub struct RecommendationsResource<'a> {
    pub(crate) client: &'a RecommendAIClient,
}

impl RecommendationsResource<'_> {
    /// Get personalised recommendations for `user_id`.
    pub fn get(&self, user_id: &str, limit: u32) -> Result<Vec<Recommendation>, Error> {
        let url = format!("{}/api/recommendations", self.client.base_url);
        let req = self.client.http.get(&url)
            .query(&[("user_id", user_id), ("limit", &limit.to_string())]);
        let resp = self.client.send(req)?;
        let data: RecommendationsResponse = resp.json()?;
        Ok(data.recommendations)
    }

    /// Get items similar to `item_id`.
    pub fn similar(&self, item_id: &str, limit: u32) -> Result<Vec<Recommendation>, Error> {
        let url = format!("{}/api/recommendations/similar/{}", self.client.base_url, item_id);
        let req = self.client.http.get(&url).query(&[("limit", &limit.to_string())]);
        let resp = self.client.send(req)?;
        let data: RecommendationsResponse = resp.json()?;
        Ok(data.recommendations)
    }

    /// Get globally popular items, optionally filtered by `category`.
    pub fn popular(&self, limit: u32, category: Option<&str>) -> Result<Vec<Recommendation>, Error> {
        let url = format!("{}/api/recommendations/popular", self.client.base_url);
        let mut params: Vec<(&str, String)> = vec![("limit", limit.to_string())];
        if let Some(cat) = category { params.push(("category", cat.to_string())); }
        let req = self.client.http.get(&url).query(&params);
        let resp = self.client.send(req)?;
        let data: RecommendationsResponse = resp.json()?;
        Ok(data.recommendations)
    }
}

// ── UsersResource ─────────────────────────────────────────────────────────────

pub struct UsersResource<'a> {
    pub(crate) client: &'a RecommendAIClient,
}

#[derive(Serialize)]
struct CreateUserBody<'a> {
    user_id:    &'a str,
    properties: &'a HashMap<String, Value>,
}

#[derive(Serialize)]
struct UpdateUserBody<'a> {
    properties: &'a HashMap<String, Value>,
}

impl UsersResource<'_> {
    pub fn create(&self, user_id: &str, properties: HashMap<String, Value>) -> Result<User, Error> {
        let url = format!("{}/api/users", self.client.base_url);
        let req = self.client.http.post(&url)
            .json(&CreateUserBody { user_id, properties: &properties });
        let resp = self.client.send(req)?;
        Ok(resp.json()?)
    }

    pub fn get(&self, user_id: &str) -> Result<User, Error> {
        let url = format!("{}/api/users/{}", self.client.base_url, user_id);
        let resp = self.client.send(self.client.http.get(&url))?;
        Ok(resp.json()?)
    }

    pub fn update(&self, user_id: &str, properties: HashMap<String, Value>) -> Result<User, Error> {
        let url = format!("{}/api/users/{}", self.client.base_url, user_id);
        let req = self.client.http.put(&url)
            .json(&UpdateUserBody { properties: &properties });
        let resp = self.client.send(req)?;
        Ok(resp.json()?)
    }

    pub fn delete(&self, user_id: &str) -> Result<(), Error> {
        let url = format!("{}/api/users/{}", self.client.base_url, user_id);
        self.client.send(self.client.http.delete(&url))?;
        Ok(())
    }
}

// ── ItemsResource ─────────────────────────────────────────────────────────────

pub struct ItemsResource<'a> {
    pub(crate) client: &'a RecommendAIClient,
}

#[derive(Serialize)]
struct CreateItemBody<'a> {
    item_id:    &'a str,
    properties: &'a HashMap<String, Value>,
}

#[derive(Serialize)]
struct UpdateItemBody<'a> {
    properties: &'a HashMap<String, Value>,
}

impl ItemsResource<'_> {
    pub fn create(&self, item_id: &str, properties: HashMap<String, Value>) -> Result<Item, Error> {
        let url = format!("{}/api/items", self.client.base_url);
        let req = self.client.http.post(&url)
            .json(&CreateItemBody { item_id, properties: &properties });
        let resp = self.client.send(req)?;
        Ok(resp.json()?)
    }

    pub fn get(&self, item_id: &str) -> Result<Item, Error> {
        let url = format!("{}/api/items/{}", self.client.base_url, item_id);
        let resp = self.client.send(self.client.http.get(&url))?;
        Ok(resp.json()?)
    }

    pub fn update(&self, item_id: &str, properties: HashMap<String, Value>) -> Result<Item, Error> {
        let url = format!("{}/api/items/{}", self.client.base_url, item_id);
        let req = self.client.http.put(&url)
            .json(&UpdateItemBody { properties: &properties });
        let resp = self.client.send(req)?;
        Ok(resp.json()?)
    }

    pub fn delete(&self, item_id: &str) -> Result<(), Error> {
        let url = format!("{}/api/items/{}", self.client.base_url, item_id);
        self.client.send(self.client.http.delete(&url))?;
        Ok(())
    }

    /// Bulk create or update items.
    pub fn upsert(&self, items: Vec<Value>) -> Result<Vec<Item>, Error> {
        let url = format!("{}/api/items/bulk", self.client.base_url);
        let body = serde_json::json!({ "items": items });
        let req  = self.client.http.post(&url).json(&body);
        let resp = self.client.send(req)?;
        let data: serde_json::Value = resp.json()?;
        let list = data["items"].as_array().cloned().unwrap_or_default();
        list.into_iter()
            .map(|v| serde_json::from_value(v).map_err(|e| Error::Http { status: 0, message: e.to_string() }))
            .collect()
    }
}

// ── InteractionsResource ──────────────────────────────────────────────────────

pub struct InteractionsResource<'a> {
    pub(crate) client: &'a RecommendAIClient,
}

#[derive(Serialize)]
struct CreateInteractionBody<'a> {
    user_id:          &'a str,
    item_id:          &'a str,
    interaction_type: &'a InteractionType,
    #[serde(skip_serializing_if = "Option::is_none")]
    value:            Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    metadata:         Option<&'a HashMap<String, Value>>,
}

impl InteractionsResource<'_> {
    pub fn create(
        &self,
        user_id:          &str,
        item_id:          &str,
        interaction_type: &InteractionType,
        value:            Option<f64>,
        metadata:         Option<&HashMap<String, Value>>,
    ) -> Result<Interaction, Error> {
        let url = format!("{}/api/interactions", self.client.base_url);
        let req = self.client.http.post(&url).json(&CreateInteractionBody {
            user_id, item_id, interaction_type, value, metadata,
        });
        let resp = self.client.send(req)?;
        Ok(resp.json()?)
    }
}
